### ISSUE TITLE
AttributeError: 'Extensions' object has no attribute 'get_extension_for_class'

### ISSUE THREAD

**Issue by matthewmok** [conda/conda#3938]
I was trying to update anaconda but have run into this error:

```
    Traceback (most recent call last):
      File "/Users/mmok/anaconda/lib/python2.7/site-packages/conda/exceptions.py", line 479, in conda_exception_handler
        return_value = func(*args, **kwargs)
      File "/Users/mmok/anaconda/lib/python2.7/site-packages/conda/cli/main.py", line 145, in _main
        exit_code = args.func(args, p)
      File "/Users/mmok/anaconda/lib/python2.7/site-packages/conda/cli/main_update.py", line 65, in execute
        install(args, parser, 'update')
      File "/Users/mmok/anaconda/lib/python2.7/site-packages/conda/cli/install.py", line 238, in install
        prefix=prefix)
      File "/Users/mmok/anaconda/lib/python2.7/site-packages/conda/api.py", line 24, in get_index
        index = fetch_index(channel_urls, use_cache=use_cache, unknown=unknown)
      File "/Users/mmok/anaconda/lib/python2.7/site-packages/conda/fetch.py", line 295, in fetch_index
        for url in urls]
      File "/Users/mmok/anaconda/lib/python2.7/site-packages/conda/fetch.py", line 75, in func
        res = f(*args, **kwargs)
      File "/Users/mmok/anaconda/lib/python2.7/site-packages/conda/fetch.py", line 117, in fetch_repodata
        timeout=(6.1, 60))
      File "/Users/mmok/anaconda/lib/python2.7/site-packages/requests/sessions.py", line 501, in get
        return self.request('GET', url, **kwargs)
      File "/Users/mmok/anaconda/lib/python2.7/site-packages/requests/sessions.py", line 488, in request
        resp = self.send(prep, **send_kwargs)
      File "/Users/mmok/anaconda/lib/python2.7/site-packages/requests/sessions.py", line 609, in send
        r = adapter.send(request, **kwargs)
      File "/Users/mmok/anaconda/lib/python2.7/site-packages/requests/adapters.py", line 423, in send
        timeout=timeout
      File "/Users/mmok/anaconda/lib/python2.7/site-packages/requests/packages/urllib3/connectionpool.py", line 594, in urlopen
        chunked=chunked)
      File "/Users/mmok/anaconda/lib/python2.7/site-packages/requests/packages/urllib3/connectionpool.py", line 350, in _make_request
        self._validate_conn(conn)
      File "/Users/mmok/anaconda/lib/python2.7/site-packages/requests/packages/urllib3/connectionpool.py", line 835, in _validate_conn
        conn.connect()
      File "/Users/mmok/anaconda/lib/python2.7/site-packages/requests/packages/urllib3/connection.py", line 330, in connect
        cert = self.sock.getpeercert()
      File "/Users/mmok/anaconda/lib/python2.7/site-packages/requests/packages/urllib3/contrib/pyopenssl.py", line 324, in getpeercert
        'subjectAltName': get_subj_alt_name(x509)
      File "/Users/mmok/anaconda/lib/python2.7/site-packages/requests/packages/urllib3/contrib/pyopenssl.py", line 171, in get_subj_alt_name
        ext = cert.extensions.get_extension_for_class(
    AttributeError: 'Extensions' object has no attribute 'get_extension_for_class'
```

Any help will be appreciated!
Thanks!

**Comment by kalefranz** [comment]
Duplicate of #3898, #3918, #3929, #3934, and #3935.  Honestly not sure where this came from all the sudden.  We definitely need to limit dependencies somewhere.

In general `pyopenssl` and `cryptography` are out of sync.

    CONDA_VERIFY_SSL=false conda update pyopenssl

should fix your problem

Do you mind posting or emailing to me your file `/Users/mmok/anaconda/conda-meta/history`.  We've got a lot of users getting into this, and I'm not sure what the exact path is.

**Comment by matthewmok** [comment]
Would that be ok if I email the file to kale@franz.io?

**Comment by kalefranz** [comment]
Yup. That'd be great. Thanks!

**Comment by kalefranz** [comment]
Leaving this issue open for now until we figure out more about it.  If you're coming here from google or otherwise

**tl;dr**

    CONDA_VERIFY_SSL=false conda update pyopenssl

**Comment by poluny** [comment]
i met a very similar problem as you , and  i find a solution
`echo $LD_LIBRARY_PATH`

**Comment by peternorwood** [comment]
i'm having this same issue when i try to create an env using "conda create --name snowflakes testprogram". typing "echo $LD_LIBRARY_PATH" doesn't help. Stuck. Appreciate any help.

**Comment by Sardtok** [comment]
I get this error on Windows. Setting the CONDA_VERIFY_SSL environment variable does not help. Is LD_LIBRARY_PATH being used on Windows? I have not setup GNU C on my system, but it shouldn't be too much of a hassle to fix. Is a similar Visual environment variable being used for linking on Windows? Cause I have Visual C++ set up.

**Comment by meereeum** [comment]
FWIW a short-term fix for this is

```
$ pip uninstall requests
$ pip install requests==2.8.1
```

(leaving behind a newer `conda` version of requests and the older `pip` version), i.e.

```
requests                  2.12.1                   py27_0  
requests                  2.8.1                     <pip>
```

This is clearly not a long-term solution though - would be good to resolve.
(`CONDA_VERIFY_SSL=false conda update pyopenssl` did nothing for me)

**Comment by kalefranz** [comment]
Sorry I'm late to respond here.  And I can't believe I was able to get this so wrong in so many places.  `CONDA_VERIFY_SSL` is *wrong*.  It must be `CONDA_SSL_VERIFY`.

    CONDA_SSL_VERIFY=false conda update pyopenssl

Also, you must be currently upgrading from conda 4.2+ for this to work.

If you have an older version of conda, use the following

    conda config --set ssl_verify false
    conda update pyopenssl
    conda config --remove ssl_verify

**Comment by meereeum** [comment]
sorry, typo above. again, I was not able to use `CONDA_SSL_VERIFY=false conda update pyopenssl` *until* i downgraded the pip `requests` library. after doing so, however, i can now upgrade `requests` with impunity!

**Comment by kalefranz** [comment]
Thanks for all the reports, help, patients, and engagement on this. Such an insidious error. It's really not conda's fault per se, but some gnarly dependency conflicts somewhere in the chain.

We have very near-term structural plans to mitigate issues like this for conda itself, as well as application dependencies in general.

**Comment by senis000** [comment]
Hi all,
I tried the above solutions and even if my conda was 4.2.13 the 
CONDA_SSL_VERIFY=false conda update pyopenssl
didn't do anything.... however the other solution:

conda config --set ssl_verify false
conda update pyopenssl

did work!!!
The third command:
conda config --remove ssl_verify
gave an error: 
conda-script.py config: error: argument --remove: expected 2 arguments

But still now I'm able to install and update anything I want :)

Thanks!

**Comment by VitamintK** [comment]
Same error as senis000, so I just assumed the proper command was `conda config --remove ssl_verify false`

**Comment by ryanriopelle** [comment]
I had a similar problem upon trying to create a new environment.  


`Current conda install:

               platform : osx-64
          conda version : 4.2.13
       conda is private : False
      conda-env version : 4.2.13
    conda-build version : 2.0.2
         python version : 3.5.2.final.0
       requests version : 2.12.4
       root environment : /Users/rriopelle-LT/anaconda3  (writable)
    default environment : /Users/rriopelle-LT/anaconda3
       envs directories : /Users/rriopelle-LT/anaconda3/envs
          package cache : /Users/rriopelle-LT/anaconda3/pkgs
           channel URLs : https://repo.continuum.io/pkgs/free/osx-64
                          https://repo.continuum.io/pkgs/free/noarch
                          https://repo.continuum.io/pkgs/pro/osx-64
                          https://repo.continuum.io/pkgs/pro/noarch
            config file : None
           offline mode : False



`$ /Users/rriopelle-LT/anaconda3/bin/conda update cffi`




    Traceback (most recent call last):
      File "/Users/rriopelle-LT/anaconda3/lib/python3.5/site-packages/conda/exceptions.py", line 479, in conda_exception_handler
        return_value = func(*args, **kwargs)
      File "/Users/rriopelle-LT/anaconda3/lib/python3.5/site-packages/conda/cli/main.py", line 145, in _main
        exit_code = args.func(args, p)
      File "/Users/rriopelle-LT/anaconda3/lib/python3.5/site-packages/conda/cli/main_update.py", line 65, in execute
        install(args, parser, 'update')
      File "/Users/rriopelle-LT/anaconda3/lib/python3.5/site-packages/conda/cli/install.py", line 238, in install
        prefix=prefix)
      File "/Users/rriopelle-LT/anaconda3/lib/python3.5/site-packages/conda/api.py", line 24, in get_index
        index = fetch_index(channel_urls, use_cache=use_cache, unknown=unknown)
      File "/Users/rriopelle-LT/anaconda3/lib/python3.5/site-packages/conda/fetch.py", line 300, in fetch_index
        repodatas = [(u, f.result()) for u, f in zip(urls, futures)]
      File "/Users/rriopelle-LT/anaconda3/lib/python3.5/site-packages/conda/fetch.py", line 300, in <listcomp>
        repodatas = [(u, f.result()) for u, f in zip(urls, futures)]
      File "/Users/rriopelle-LT/anaconda3/lib/python3.5/concurrent/futures/_base.py", line 405, in result
        return self.__get_result()
      File "/Users/rriopelle-LT/anaconda3/lib/python3.5/concurrent/futures/_base.py", line 357, in __get_result
        raise self._exception
      File "/Users/rriopelle-LT/anaconda3/lib/python3.5/concurrent/futures/thread.py", line 55, in run
        result = self.fn(*self.args, **self.kwargs)
      File "/Users/rriopelle-LT/anaconda3/lib/python3.5/site-packages/conda/fetch.py", line 75, in func
        res = f(*args, **kwargs)
      File "/Users/rriopelle-LT/anaconda3/lib/python3.5/site-packages/conda/fetch.py", line 117, in fetch_repodata
        timeout=(6.1, 60))
      File "/Users/rriopelle-LT/anaconda3/lib/python3.5/site-packages/requests/sessions.py", line 501, in get
        return self.request('GET', url, **kwargs)
      File "/Users/rriopelle-LT/anaconda3/lib/python3.5/site-packages/requests/sessions.py", line 488, in request
        resp = self.send(prep, **send_kwargs)
      File "/Users/rriopelle-LT/anaconda3/lib/python3.5/site-packages/requests/sessions.py", line 609, in send
        r = adapter.send(request, **kwargs)
      File "/Users/rriopelle-LT/anaconda3/lib/python3.5/site-packages/requests/adapters.py", line 423, in send
        timeout=timeout
      File "/Users/rriopelle-LT/anaconda3/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py", line 594, in urlopen
        chunked=chunked)
      File "/Users/rriopelle-LT/anaconda3/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py", line 350, in _make_request
        self._validate_conn(conn)
      File "/Users/rriopelle-LT/anaconda3/lib/python3.5/site-packages/requests/packages/urllib3/connectionpool.py", line 835, in _validate_conn
        conn.connect()
      File "/Users/rriopelle-LT/anaconda3/lib/python3.5/site-packages/requests/packages/urllib3/connection.py", line 330, in connect
        cert = self.sock.getpeercert()
      File "/Users/rriopelle-LT/anaconda3/lib/python3.5/site-packages/requests/packages/urllib3/contrib/pyopenssl.py", line 324, in getpeercert
        'subjectAltName': get_subj_alt_name(x509)
      File "/Users/rriopelle-LT/anaconda3/lib/python3.5/site-packages/requests/packages/urllib3/contrib/pyopenssl.py", line 171, in get_subj_alt_name
        ext = cert.extensions.get_extension_for_class(
    AttributeError: 'Extensions' object has no attribute 'get_extension_for_class'
`

**Comment by kalefranz** [comment]
Solutions posted above.  Root cause is dependency mismatch between requests and it's own dependencies.
